rfcs/rfc-001-smart-contracts-engine.md

```diff
@@ -37,154 +37,175 @@ version 0.1
+### State
 
+State is addressed by an account which is at the moment simply the PubKey.  Our goal is to eliminate dynamic memory allocation in the smart contract itself, so the contract is a function that takes a mapping of [(PubKey,State)] and returns [(PubKey, State')].
 
-### Call
-
+### Call Structure
 ```
-void call(
-    const struct instance_data *data,
-    const uint8_t kind[],  //instance|participant|caller|read|write
-    const uint8_t *keys[],
-    uint8_t *data[],
-    int num,
-    uint8_t dirty[],        //dirty memory bits
-    uint8_t *userdata,      //current transaction data
-);
+/// Call definition
+/// Signed portion
+#[derive(Serialize, Deserialize, Debug, PartialEq, Eq, Clone)]
+pub struct CallData {
+    /// number of keys to load, aka the to key
+    /// keys[0] is the caller's key
+    pub keys: Vec<PublicKey>,
+
+    /// the set of PublicKeys that are required to have a proof
+    pub required_proofs: Vec<u8>,
+
+    /// PoH data
+    /// last id PoH observed by the sender
+    pub last_count: u64,
+    /// last PoH hash observed by the sender
+    pub last_hash: Hash,
+
+    /// Program
+    /// the address of the program we want to call
+    pub contract: ContractId,
+    /// OS scheduling fee
+    pub fee: i64,
+    /// struct version to prevent duplicate spends
+    /// Calls with a version <= Page.version are rejected
+    pub version: u64,
+    /// method to call in the contract
+    pub method: u8,
+    /// usedata in bytes
+    pub user_data: Vec<u8>,
+}
+
+#[derive(Serialize, Deserialize, Debug, PartialEq, Eq, Clone)]
+pub struct Call {
+    /// Signatures and Keys
+    /// (signature, key index)
+    /// This vector contains a tuple of signatures, and the key index the signature is for
+    /// proofs[0] is always key[0]
+    pub proofs: Vec<Signature>,
+    pub data: CallData,
+}
 ```
 
-To call this operation, the transaction that is destined to the contract instance specifies what keyed state it should present to the `call` function.  To allocate the state memory or a call context, the client has to first call a function on the contract with the designed address that will own the state.
-
-At its core, this is a system call that requires cryptographic proof of ownership of memory regions instead of an OS that checks page tables for access rights.
-
-* `Instance_AllocateContext(Instance PubKey, My PubKey, Proof of key ownership)`
-
-Any transaction can then call `call` on the contract with a set of keys.  It's up to the contract itself to manage ownership:
-
-* `Instance_Call(Instance PubKey, [Context PubKeys], proofs of ownership, userdata...)`
+At it's core, this is just a set of PublicKeys and Signatures with a bit of metadata.  The contract PublicKey routes this transaction into that contracts entry point.  `version` is used for dropping retransmitted requests.
```

Comment:
its. contract's.

Should `version` be boot from this PR, or is it still relevant in this context?